pkg/controller/argocd/register_test.go

```diff
@@ -0,0 +1,111 @@
+	switch o := v.(type) {
+	case *appsv1.Deployment:
+		var replicas int32 = 3
+		o.Spec.Replicas = &replicas
+	}
+	return nil
+}
+
+func testClusterRoleHook(cr *argoprojv1alpha1.ArgoCD, v interface{}) error {
+	switch o := v.(type) {
+	case *v1.ClusterRole:
+		o.Rules = append(o.Rules, policyRuleForApplicationController()...)
+	}
+	return nil
+}
+
+func testErrorHook(cr *argoprojv1alpha1.ArgoCD, v interface{}) error {
+	switch o := v.(type) {
+	case *appsv1.Deployment:
+		err := fmt.Errorf("Error returned from %s for %s", "testErrorHook", o.Name)
+		return err
+	}
+	return nil
+}
+
+func TestReconcileArgoCD_testDeploymentHook(t *testing.T) {
+	defer stubDefaultHookFunc(t)()
+	logf.SetLogger(logf.ZapLogger(true))
+	a := makeTestArgoCDForClusterConfig()
+
+	Register(testDeploymentHook)
+
+	testDeployment := makeTestDeployment()
+	if err := applyReconcilerHook(a, testDeployment); err != nil {
+		t.Fatalf("failed to apply reconciler hook: %#v\n", err)
+	}
+	var expectedReplicas int32 = 3
+	assert.DeepEqual(t, &expectedReplicas, testDeployment.Spec.Replicas)
+}
+
+func TestReconcileArgoCD_testMultipleHooks(t *testing.T) {
+	defer stubDefaultHookFunc(t)()
+	logf.SetLogger(logf.ZapLogger(true))
+	a := makeTestArgoCDForClusterConfig()
+
+	testDeployment := makeTestDeployment()
+	testClusterRole := makeTestClusterRole()
+
+	Register(testDeploymentHook)
+	Register(testClusterRoleHook)
+
+	if err := applyReconcilerHook(a, testDeployment); err != nil {
+		t.Fatalf("failed to apply reconciler hook: %#v\n", err)
+	}
+
+	if err := applyReconcilerHook(a, testClusterRole); err != nil {
+		t.Fatalf("failed to apply reconciler hook: %#v\n", err)
+	}
+
+	// Verify if testDeploymentHook is executed successfully
+	var expectedReplicas int32 = 3
+	assert.DeepEqual(t, &expectedReplicas, testDeployment.Spec.Replicas)
+
+	// Verify if testClusterRoleHook is executed successfully
+	want := append(makeTestPolicyRules(), policyRuleForApplicationController()...)
+	assert.DeepEqual(t, want, testClusterRole.Rules)
+}
+
+func TestReconcileArgoCD_testErrorHook(t *testing.T) {
+	defer stubDefaultHookFunc(t)()
+	logf.SetLogger(logf.ZapLogger(true))
+	a := makeTestArgoCDForClusterConfig()
+
+	// Register a hook that errors for a Deployment Resource
+	Register(testErrorHook)
+
+	// Error Hook should not effect ClusterRole Resource
+	testClusterRole := makeTestClusterRole()
+	if err := applyReconcilerHook(a, testClusterRole); err != nil {
+		t.Fatalf("failed to apply reconciler hook: %#v\n", err)
+	}
+	assert.DeepEqual(t, makeTestPolicyRules(), testClusterRole.Rules)
+
+	// Error should effect the Deployment Resource
+	testDeployment := makeTestDeployment()
+	err := applyReconcilerHook(a, testDeployment)
+	expectedError := fmt.Sprintf("Error returned from %s for %s", "testErrorHook", testDeployment.Name)
+	assert.Error(t, err, expectedError, applyReconcilerHook(a, testDeployment))
+}
+
+func stubDefaultHookFunc(t *testing.T) func() {
+	origDefaultHooksFunc := hooks
```

Comment:
You're not using `t` in this function?

Also, how about we name this `resetHooks` ?

I'd also be inclined to do something like `hooks = []Hook{}` which would mean that you have a known state for the tests`